Fixed HostFilteringMiddleware. Added HostFilteringMiddleware.MiddlewareConfigurationManager

charytech opened the pull request:
# Fixed HostFilteringMiddleware. Added HostFilteringMiddleware.MiddlewareConfigurationManager

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

## Description

Added HostFilteringMiddleware.MiddlewareConfigurationManager which is responisible for maintenence configuration state used by middleware invocation.

Fixes #51960

charytech committed:
Fixed HostFilteringMiddleware. Added HostFilteringMiddleware.MiddlewareConfigurationManager

charytech committed:
update  logic

charytech committed:
Added extra logs

charytech commented:
@dotnet-policy-service agree

charytech committed:
Simplify conditions

charytech committed:
reverted original comment

charytech committed:
Formating

Tratcher commented on src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs:
```diff
@@ -62,140 +55,196 @@ public class HostFilteringMiddleware
+            ); // IPv4 Any
+        }
+
+        private sealed record MiddlewareHostFilteringOptions(ReadOnlyCollection<string> AllowedHosts, bool AllowEmptyHosts, bool IncludeFailureMessage)
```
I'm not sure why you need this type either? HostFilteringOptions returned from IOptionsMonitor should be stable.

charytech replied:
HostFilteringOptions provided from IOptionsMonitor Onchange method is a reference type and it could be changed in any other event handlers, so we should avoid such situations, to achieve this we can create a copy of HostFilteringOptions or use our internal type during evaluation.

Also, the new type could be useful for checking if the new values will be different from the previous ones, and then we can skip the configuration re-evaluation. But it is a good candidate for the future and is not covered by this bug fix.

Tratcher replied:
Maybe in theory, but this is the only place this type is consumed so such defense is not needed. We'd rather avoid the extra allocations and work. We'll address it if it becomes an issue in the future.

charytech replied:
I understand your point as it is really an edge case. But why are you afraid of simple allocation? If users will not update configs often it will be only single allocation. But two quick update can cause various strange behaviors which will be difficult to debug. I can see quite clear benefits of doing it in atomic way and code is more stable.
What do you think?

Tratcher replied:
Keep It Simple. Don't solve theoretical problems, only real ones. Unnecessary code makes components harder to understand and maintain.

charytech replied:
Ok. I will change it as you want

charytech commented on src/Middleware/HostFiltering/src/HostFilteringMiddleware.cs:
```diff
@@ -62,140 +55,196 @@ public class HostFilteringMiddleware
+
+    internal sealed record MiddlewareConfiguration(bool AllowAnyNonEmptyHost, bool AllowEmptyHosts, bool IncludeFailureMessage, ReadOnlyCollection<StringSegment>? AllowedHosts = default);
+
+    internal sealed class MiddlewareConfigurationManager
```
Hey, 
1. Separation of responsibility. 
2. I think it is easier to maintain code.
3. We can always add everything under one object but it looks more clear to me.

Tratcher replied:
It's substantial enough that it deserves to be in its own file. Nesting large components makes things harder to navigate and make sense of.

charytech committed:
Changes related to pull request review, MiddlewareConfigurationManager moved to own file, removed MiddlewareHostFilteringOptions object.

charytech committed:
ConfigureMiddleware atomic

charytech commented:
Hey, @Tratcher I changed it as you wanted, removed an extra object, and moved Manager to a new file.
I left a new object allocation because if we don't want to do it lock statement does not make sense. It will be allocated only when a change request is registered. 
Config can change during the lifecycle of Middleware so it is expected behavior as we are using IOptionsManager, but config shouldn't change during ConfigureMiddleware and it should be an atomic operation. 
Without it could bring us unexpected behavior, code should be stable.

charytech committed:
Revert "ConfigureMiddleware atomic"

This reverts commit 661e6f608d0168b4c30ceca292d0b843aac1a252.

charytech commented:
Hey @Tratcher sorry for the delay. Changed as you suggested.

charytech commented:
what about previous versions? will you patch them?

Tratcher commented on src/Middleware/HostFiltering/src/MiddlewareConfigurationManager.cs:
```diff
@@ -0,0 +1,120 @@
+        _getChangeRequestObject = () => _optionsMonitor.CurrentValue;
+        _optionsMonitor.OnChange(options =>
+        {
+            lock (_syncObject)
+            {
+                _getChangeRequestObject = () => options;
+                if (_logger.IsEnabled(LogLevel.Debug))
+                {
+                    _logger.OptionsChanged(GetLogMessageForHostFilteringOptions(options));
+                }
+            }
```
```suggestion
            _middlewareConfiguration = ConfigureMiddleware(options);
```

Tratcher commented on src/Middleware/HostFiltering/src/MiddlewareConfigurationManager.cs:
```diff
@@ -0,0 +1,120 @@
+    {
+        _logger = logger;
+        // Configuration will evaluate during first request
+        _getChangeRequestObject = () => _optionsMonitor.CurrentValue;
```
```suggestion
        _middlewareConfiguration = ConfigureMiddleware(options);
```

Tratcher commented on src/Middleware/HostFiltering/src/MiddlewareConfigurationManager.cs:
```diff
@@ -0,0 +1,120 @@
+{
+    private readonly object _syncObject = new();
+    private event Func<HostFilteringOptions>? _getChangeRequestObject;
+    private MiddlewareConfiguration? _middlewareConfiguration;
```
```suggestion
    private MiddlewareConfiguration _middlewareConfiguration;
```

Tratcher commented on src/Middleware/HostFiltering/src/MiddlewareConfigurationManager.cs:
```diff
@@ -0,0 +1,120 @@
+
+internal sealed class MiddlewareConfigurationManager
+{
+    private readonly object _syncObject = new();
+    private event Func<HostFilteringOptions>? _getChangeRequestObject;
```
I don't think this locking or indirection are needed. We just need to make sure _middlewareConfiguration is handled atomically.
```suggestion
```

Tratcher commented on src/Middleware/HostFiltering/src/MiddlewareConfigurationManager.cs:
```diff
@@ -0,0 +1,120 @@
+    }
+    internal MiddlewareConfiguration GetLatestMiddlewareConfiguration()
+    {
+        if (_getChangeRequestObject is not null)
+        {
+            HostFilteringOptions options;
+            lock (_syncObject)
+            {
+                options = _getChangeRequestObject();
+                _getChangeRequestObject = null;
+            }
+            _middlewareConfiguration = ConfigureMiddleware(options);
+        }
+
+        if (_middlewareConfiguration?.AllowAnyNonEmptyHost != true && (_middlewareConfiguration?.AllowedHosts is null || _middlewareConfiguration.AllowedHosts?.Count == 0))
```
```suggestion
        var config = _middlewareConfiguration;
        if (!config.AllowAnyNonEmptyHost && (config.AllowedHosts is null || config.AllowedHosts.Count == 0))
```

You can also add this check in the constructor.

charytech replied:
Why do you want to do all operations in the constructor?
_Constructors should not do much work other than capture the constructor parameters. The cost of any other processing should be delayed until required._

The configuration of the middleware should be evaluated as needed. 
Even from a performance perspective, it's better.
You can have several HostFilteringOptions updates (e.g. 5 ) but only one request after those updates so reconfiguration should be done only once.
That's why I created the delegate, to transfer responsibility from the constructor to the method when configuration is needed.

Tratcher replied:
Constructors should make sure the object is initialized and ready to use. This data is going to be needed on all requests, so it makes sense to get it ready in advance. It's not a lot of extra work in this case.

> You can have several HostFilteringOptions updates (e.g. 5 ) but only one request after those updates so reconfiguration should be done only once.

Requests are _much_ more common than options updates.

charytech replied:
> Constructors should make sure the object is initialized and ready to use. This data is going to be needed on all requests, so it makes sense to get it ready in advance. It's not a lot of extra work in this case.

why my implementation is wrong? it is doing the same as you mentioned it is initializing data to be ready to use, with the difference that middleware initialization is faster and reevaluation is done only when needed, you can have a bigger list of allowed hosts that will use not needed resources to reevaluate state.

> Requests are _much_ more common than options updates.
You are right but it is theoretical

Tratcher replied:
My main complaint is that the implementation is more complex than necessary. Keep it simple.

As for perf, avoid doing work on a request thread when possible, do it at startup or on an event thread instead.

charytech replied:
I can partially agree as it is better to do work on an event thread :D but about complexity, I think it is still simple 10 extra lines with logging. 
We have two different concepts which have their own pros and cons.
I kept the current implementation because the configuration was evaluated during the request, I also agree with this because it is better to avoid unnecessary evaluation. Re-evaluation on the event thread is slowing down event processing, any nested methods assigned to an event will be delayed. 

You maintain this code so I will change it as you requested :)

Tratcher commented on src/Middleware/HostFiltering/src/MiddlewareConfigurationManager.cs:
```diff
@@ -0,0 +1,120 @@
+            _logger.MiddlewareConfigurationStarted(GetLogMessageForHostFilteringOptions(options));
+        }
+
+        if (
+            options.AllowedHosts?.Count > 0
+            && !TryProcessHosts(options.AllowedHosts, allowedHosts)
+        )
```
```suggestion
        if (options.AllowedHosts?.Count > 0
            && !TryProcessHosts(options.AllowedHosts, allowedHosts))
```

Tratcher commented on src/Middleware/HostFiltering/src/MiddlewareConfigurationManager.cs:
```diff
@@ -0,0 +1,120 @@
+
+    private static bool IsTopLevelWildcard(string host)
+    {
+        return (
+            string.Equals("*", host, StringComparison.Ordinal) // HttpSys wildcard
+            || string.Equals("[::]", host, StringComparison.Ordinal) // Kestrel wildcard, IPv6 Any
+            || string.Equals("0.0.0.0", host, StringComparison.Ordinal)
+        ); // IPv4 Any
```
```suggestion
        return string.Equals("*", host, StringComparison.Ordinal) // HttpSys wildcard
            || string.Equals("[::]", host, StringComparison.Ordinal) // Kestrel wildcard, IPv6 Any
            || string.Equals("0.0.0.0", host, StringComparison.Ordinal); // IPv4 Any
```

Tratcher commented on src/Middleware/HostFiltering/src/MiddlewareConfigurationManager.cs:
```diff
@@ -0,0 +1,120 @@
+            || string.Equals("[::]", host, StringComparison.Ordinal) // Kestrel wildcard, IPv6 Any
+            || string.Equals("0.0.0.0", host, StringComparison.Ordinal)
+        ); // IPv4 Any
+    }
+    private static string GetLogMessageForHostFilteringOptions(HostFilteringOptions hostFilteringOptions)
```
```suggestion
    }
    
    private static string GetLogMessageForHostFilteringOptions(HostFilteringOptions hostFilteringOptions)
```

charytech commented:
@Tratcher  Why do you want to do all operations in the constructor?
Constructors should not do much work other than capture the constructor parameters. The cost of any other processing should be delayed until required.

The configuration of the middleware should be evaluated as needed.
Even from a performance perspective, it's better.
You can have several HostFilteringOptions updates (e.g. 5 ) but only one request after those updates so reconfiguration should be done only once.
That's why I created the delegate, to transfer responsibility from the constructor to the method when configuration is needed.

charytech committed:
PR comments

charytech committed:
format

charytech committed:
format

charytech commented:
@Tratcher  I think I resolved all of your comments

Tratcher commented:
Thanks!
